lib/Partitioner/PartitionerUtils.cpp

```diff
@@ -140,72 +140,154 @@ getOutUsersWithOnePredecessor(const std::set<Node *> &nodes) {
   return ret;
 }
 
+/// \returns the list of users who use \p node's \p idx - th output.
+std::vector<Node *> getNodeUsers(Node *node, size_t idx) {
+  std::vector<Node *> ret;
+  NodeValue nodeVal = node->getNthResult(idx);
+  for (auto &U : node->getUsers()) {
+    Node *user = U.getUser();
+    for (size_t i = 0, e = user->getNumInputs(); i < e; i++) {
+      NodeValue val = user->getNthInput(i);
+      if (val == nodeVal) {
+        ret.push_back(user);
+        break;
+      }
+    }
+  }
+  return ret;
+}
+
+/// \returns the memory usage of the output caused by \p node who has users not
+/// in the set \p nodes.
 uint64_t getOutMemPerNode(const std::set<Node *> &nodes, const Node *node) {
   uint64_t ret = 0;
-  // Assume that for each Node, the max number of outputs is 5 so far.
-  llvm::SmallSet<size_t, 5> used;
-  for (const auto &U : node->getUsers()) {
-    const Node *user = U.getUser();
-    if (nodes.find(const_cast<Node *>(user)) != nodes.end()) {
-      continue;
-    }
-    // Find the user that doesn't belong to this subgraph.
-    // If a node has several users, the memory usage is only counted once.
-    for (size_t idx = 0, end = user->getNumInputs(); idx < end; idx++) {
-      NodeValue val = user->getNthInput(idx);
-      if (val.getNode() == node && used.count(val.getResNo()) == 0) {
-        ret += node->getType(val.getResNo())->getSizeInBytes();
-        used.insert(val.getResNo());
+  for (size_t i = 0, e = node->getNumResults(); i < e; i++) {
+    std::vector<Node *> users = getNodeUsers(const_cast<Node *>(node), i);
+    for (auto &user : users) {
+      if (nodes.find(const_cast<Node *>(user)) == nodes.end()) {
+        ret += node->getType(i)->getSizeInBytes();
+        break;
       }
     }
   }
   return ret;
 }
 
-GraphMemInfo getGraphMemInfo(const std::set<Node *> &nodes) {
-  GraphMemInfo ret;
-  std::set<NodeValue> nSet;
-  for (std::set<Node *>::iterator it = nodes.begin(); it != nodes.end(); ++it) {
-    Node *cur = *it;
-    // For Save onde, the only required memory is for output.
-    if (auto *SN = llvm::dyn_cast<SaveNode>(cur)) {
-      Storage *out = llvm::dyn_cast<Storage>(SN->getOutput().getNode());
-      ret.outMemSize += out->getType()->getSizeInBytes();
+/// Given nodes set \p nodes and its memory usage info \p info, \returns the new
+/// memory usage if \p node is added into \p nodes.
+GraphMemInfo updateGraphMemInfoByAddingNode(const std::set<Node *> &nodes,
+                                            const GraphMemInfo &info,
+                                            Node *node) {
+  GraphMemInfo ret = info;
```

Comment:
I moved next line after the first loop. But for the "GraphMemInfo ret = info;" I would like to keep it as the first line, since it is the return value of this function :)

Reply:
I moved next line after the first loop. But for the "GraphMemInfo ret = info;" I would like to keep it as the first line, since it is the return value of this function :)

Reply:
Everywhere this function is called, after we call it with some `GraphMemInfo info` we never use the `info` that was passed in again, right? So it seems unnecessary to create a copy, no?

Reply:
No. It is not actually "adding", just a prediction at this stage.